lsp/lib/prover/toolbox.ml

```diff
@@ -0,0 +1,429 @@
+type tlapm_obl_state =
+  | ToBeProved
+  | BeingProved
+  | Normalized
+  | Proved
+  | Failed
+  | Interrupted
+  | Trivial
+  | Unknown of string
+
+let tlapm_obl_state_of_string s =
+  match s with
+  | "to be proved" -> ToBeProved
+  | "being proved" -> BeingProved
+  | "normalized" -> Normalized
+  | "proved" -> Proved
+  | "failed" -> Failed
+  | "interrupted" -> Interrupted
+  | "trivial" -> Trivial
+  | _ -> Unknown s
+
+(* The strings here don't have to match with the ones above. *)
+let tlapm_obl_state_to_string (s : tlapm_obl_state) : string =
+  match s with
+  | ToBeProved -> "to be proved"
+  | BeingProved -> "being proved"
+  | Normalized -> "normalized"
+  | Proved -> "proved"
+  | Failed -> "failed"
+  | Interrupted -> "interrupted"
+  | Trivial -> "trivial"
+  | Unknown s -> "unknown state: " ^ s
+
+module Obligation = struct
+  type t = {
+    id : int;
+    loc : Range.t;
+    status : tlapm_obl_state;
+    fp : string option;
+    prover : string option;
+    meth : string option;
+    reason : string option;
+    already : bool option;
+    obl : string option;
+  }
+
+  (** Indicates if we should not expect any other obligation messages with the same id. *)
+  let is_final o =
+    match o.status with
+    | ToBeProved -> false
+    | BeingProved -> false
+    | Normalized -> false
+    | Proved -> true
+    | Failed -> ( match o.prover with Some "isabelle" -> true | _ -> false)
```

Comment:
Thanks for pointing this out. I don't see how a client (LSP in this case) can determine if all attempts on the obligation are done. Can I extend the toolbox protocol on this by adding a list of pending provers in the initial message for the obligation? It would also be nice for the IDE user to see a list of pending provers for a step from the beginning. E.g.:

```
@!!type:obligation
@!!id:1
@!!loc:4:7:4:9
@!!status:to be proved
@!!provers:isabelle,z3 <------- added as a proposal.
@!!END
```

Reply:
That's a good idea, but we'll have to check that the toolbox itself properly ignores this new field.
@lemmy should be able to tell us.

Reply:
I wasn't involved in designing the TLAPS <> Toolbox protocol. I can start investigating, but I suggest we ask @xxyzzn at the next TLA+ Community meeting if there was a specific reason for not including this in the protocol originally.

Reply:
Resolved in 8138f4bc6ac1b65015bc82fb7871677b26dd467b.
A switch `--toolbox-vsn` was added, which defaults to `1` -- the old version.
In the new version (`2`) a message like the following is printed additionally:
```
@!!type:obligationprovers
@!!id:1
@!!provers:smt,zenon,isabelle
@!!END
```

On the LSP side I handle this message and check the finality of an obligation's
state without relying on the order of the provers.